ssl/quic/quic_txp.c

```diff
@@ -0,0 +1,2027 @@
+
+    cc_limit_ = txp->args.cc_method->get_send_allowance(txp->args.cc_data,
+                                                        time_since_last,
+                                                        ossl_time_is_zero(time_since_last));
+
+    cc_limit = (cc_limit_ > SIZE_MAX ? SIZE_MAX : (size_t)cc_limit_);
+
+    /* Assemble packet header. */
+    phdr.type           = ossl_quic_enc_level_to_pkt_type(enc_level);
+    phdr.spin_bit       = 0;
+    phdr.pn_len         = txp_determine_pn_len(txp);
+    phdr.partial        = 0;
+    phdr.fixed          = 1;
+    phdr.version        = QUIC_VERSION_1;
+    phdr.dst_conn_id    = txp->args.cur_dcid;
+    phdr.src_conn_id    = txp->args.cur_scid;
+
+    /*
+     * We need to know the length of the payload to get an accurate header
+     * length for non-1RTT packets, because the Length field found in
+     * Initial/Handshake/0-RTT packets uses a variable-length encoding. However,
+     * we don't have a good idea of the length of our payload, because the
+     * length of the payload depends on the room in the datagram after fitting
+     * the header, which depends on the size of the header.
+     *
+     * In general, it does not matter if a packet is slightly shorter (because
+     * e.g. we predicted use of a 2-byte length field, but ended up only needing
+     * a 1-byte length field). However this does matter for Initial packets
+     * which must be at least 1200 bytes, which is also the assumed default MTU;
+     * therefore in many cases Initial packets will be padded to 1200 bytes,
+     * which means if we overestimated the header size, we will be short by a
+     * few bytes and the server will ignore the packet for being too short. In
+     * this case, however, such packets always *will* be padded to meet 1200
+     * bytes, which requires a 2-byte length field, so we don't actually need to
+     * worry about this. Thus we estimate the header length assuming a 2-byte
+     * length field here, which should in practice work well in all cases.
+     */
+    phdr.len            = OSSL_QUIC_VLINT_2B_MAX - phdr.pn_len;
+
+    if (enc_level == QUIC_ENC_LEVEL_INITIAL) {
+        phdr.token      = txp->initial_token;
+        phdr.token_len  = txp->initial_token_len;
+    } else {
+        phdr.token      = NULL;
+        phdr.token_len  = 0;
+    }
+
+    hdr_len = ossl_quic_wire_get_encoded_pkt_hdr_len(phdr.dst_conn_id.id_len,
+                                                     &phdr);
+    if (hdr_len == 0)
+        return TXP_ERR_INPUT;
+
+    /* MinDPL: Minimum total datagram payload length. */
+    min_dpl = must_pad ? QUIC_MIN_INITIAL_DGRAM_LEN : 0;
+
+    /* How much data is already in the current datagram? */
+    running_total = ossl_qtx_get_cur_dgram_len_bytes(txp->args.qtx);
+
+    /* MinPL: Minimum length of the fully encoded packet. */
+    min_pl = running_total < min_dpl ? min_dpl - running_total : 0;
+    if ((uint64_t)min_pl > cc_limit)
```

Comment:
I see no reason why congestion control can't apply to an Initial packet, unless I've forgotten something in the RFCs. In the practical case Initial is done with pretty quickly but abstractly this needn't be the case; QUIC has fairly clean separation between the bulk of it and the handshake protocol and theoretically could even be used with a different handshake protocol involving many round trips in Initial.

In any case it's moot, since padding isn't only relevant to Initial packets. We will also need it when we eventually implement MTU probing, and will probably want to allow applications to configure extra padding for anti-traffic-analysis purposes. So this isn't Initial-specific.